### PR TITLE
array pool refactor

### PR DESCRIPTION
**Pull request by YunHsiao**
Re: cocos-creator/3d-tasks#4535

Changelog:
 * we need some kind of raw buffer pool for things like instance buffers. refactored array pool to extends base class `BufferAllocator`, which is designed to allocate exactly that.

### CODE REVIEW

File: cocos/core/renderer/scene/render-scene.ts [hunk 1]
```diff
@@ -159,9 +159,11 @@ export class RenderScene {
         this.removeSpotLights();
         this.removeModels();
         if (this._modelArrayHandle) {
-            ModelArrayPool.free(this._modelArrayHandle);
-            ScenePool.free(this._scenePoolHandle);
+            freeHandleArray(this._modelArrayHandle, ModelArrayPool, ModelPool);
```
Comment (minggo) [inline cocos/core/renderer/scene/render-scene.ts]: `this.removeModels();` 会调用 `Model.destroy()`，从而释放 Model 自己的 handle。为什么需要再次释放呢？

File: cocos/core/renderer/ui/ui-batch-model.ts [hunk 2]
```diff
@@ -67,7 +67,7 @@ export class UIBatchModel extends Model {
 
     public directInitialize (batch: UIDrawBatch) {
         this._subModel.directInitialize(batch.material!.passes, batch.hInputAssembler, batch.hDescriptorSet!);
-        SubModelArrayPool.assign(this._subModelArrayHandle, 0, this._subModel.handle);
+        SubModelArrayPool.assign(ModelPool.get(this._handle, ModelView.SUB_MODEL_ARRAY), 0, this._subModel.handle);
```
Comment (minggo) [inline cocos/core/renderer/ui/ui-batch-model.ts]: _subModelArrayHandle 直接从父类继承下来，没必要再通过 ModelPool 去获取吧？

File: cocos/core/renderer/scene/render-scene.ts (hunk 1 above)
Comment (YunHsiao) [reply cocos/core/renderer/scene/render-scene.ts]: 噢这个不应该改

File: cocos/core/renderer/ui/ui-batch-model.ts (hunk 2 above)
Comment (YunHsiao) [reply cocos/core/renderer/ui/ui-batch-model.ts]: 没看到这种高频调用的地方还用到了，我把父类里的成员变量删了，我改下子类吧

**Pull request merged**